### ISSUE TITLE
Use TypeScript 5.4

### ISSUE THREAD

**Issue by lcswillems** [pulumi/pulumi#12117]
Pulumi uses TypeScript 3.8.3 released on February 20, 2020. Latest Typescript is 5.4 released on March 6, 2024.

3.8.3 is 4 years old! A ton of TypeScript features are missing.

**Comment by justinvp** [comment]
Potentially related to https://github.com/pulumi/pulumi/issues/8177 (making typescript and ts-node peer dependencies)

There's been a few things holding us back from upgrading the built-in TypeScript version:
- Compatibility for existing projects
- ts-node performance
- Tooling we use to generate SDK docs from the TypeScript source

We should come up with a solution as we really should aim to be able to support the latest language features out of the box.


## Workaround

In the meantime, it is possible to use the latest versions of TypeScript with Pulumi, it just requires you to compile your TypeScript (or any language that transpiles to JavaScript) ahead of time. You can also use whatever tooling you like, like SWC.

From our [docs](https://www.pulumi.com/docs/intro/languages/javascript/#typescript):

> If you would like full control of the TypeScript build process, you can compile ahead of time, and point your package.json main entry point at the compiled JavaScript instead. If you do this, you can disable the [automatic compilation of TypeScript files](https://www.pulumi.com/docs/intro/languages/javascript/#disabling-built-in-typescript-support).

And:

> You can disable the built in TypeScript support by changing the `runtime` setting in `Pulumi.yaml` to look like the following:
> ```yaml
> runtime:
>   name: nodejs
>   options:
>     typescript: false
> ```

**Comment by lcswillems** [comment]
My solution for the moment (the simplest one) is to use package.json resolutions and overwrite the version of TypeScript used:

```
  "resolutions": {
    "typescript": "4.9.5"
  }
```

It would also be nice to be able to use .js without having to disable TypeScript, e.g. by putting `main: "index.js"` in package.json. If extension is js, then it is Pulumi JS and if extension is ts, then it is Pulumi TS that runs.

**Comment by RobbieMcKinstry** [comment]
> It would also be nice to be able to use .js without having to disable TypeScript, e.g. by putting main: "index.js" in package.json. If extension is js, then it is Pulumi JS and if extension is ts, then it is Pulumi TS that runs.

Hm, I thought we implemented that last year, but perhaps I was mistaken. @lcswillems would you mind opening an issue?

**Comment by valkum** [comment]
It would be nice if you could define a command that is run before Pulumi looks at the `main` file defined in the package.json.
```yaml
runtime:
  name: nodejs
  options:
    typescript: false
    prepare: yarn run compile
```
This avoids the need to manually call the compile step ahead of time. I think this would be a good middle ground between keeping a stable, backwards-compatible TypeScript version pinned in Pulumi and the option to manage the ahead of time compilation to JS yourself.

**Comment by JonParton** [comment]
I also hit this today trying to use `satisfies` 😔

Any news on the pulumi typescript version being updated and keeping pace with the power TS devs will be used to wielding!?! 🙏

**Comment by RobbieMcKinstry** [comment]
Hi @JonParton ,

I'm no longer an OSS maintainer with Pulumi, so ownership of this work has been transferred back to the Pulumi Core team. Wishing you all the best with Pulumi and TypeScript! 🚀

**Comment by lcswillems** [comment]
@justinvp Any news?

I have also opened another issue to support Bun 1.0 runtime: https://github.com/pulumi/pulumi/issues/13904

Maybe both could be fixed at the same time.

**Comment by justinvp** [comment]
No news other than I'm hoping we can get to this soon. Thanks for opening https://github.com/pulumi/pulumi/issues/13904

**Comment by medv** [comment]
> 3. Include a configuration option in `Pulumi.yaml` that allows you to provide your own compilation workflow.

This, as well as analogous cli arguments would seem to be the best way to go. Good composition gets us further and removes the most complaints.

**Comment by remagpie** [comment]
For those looking for a workaround to use your own version of typescript, this recipe worked for my case.

1. Add typescript and ts-node on your package.json. Change the package versions as you want.
```json
  "dependencies": {
    "@pulumi/pulumi": "^3.0.0",
    "ts-node": "^10.9.2",
    "typescript": "^5.3.3"
  }
```
2. Disable the built in typescript support and register ts-node in Pulumi.yaml
```yaml
name: Some name
description: Some description
runtime:
  name: nodejs
  options:
    typescript: false
    nodeargs: "-r ts-node/register"
```

Now scripts like `const foo = {} satisfies Record<string, string>` runs without an error.

**Comment by mattfysh** [comment]
if you want to avoid compiling to commonjs, and have your program executed as esm modules (important for some language features), `ts-node` is not going to be a good fit. You might try tsx using @remagpie snippet above with this value: `nodeargs: --import tsx/esm`

**Comment by medv** [comment]
> if you want to avoid compiling to commonjs, and have your program executed as esm modules (important for some language features), `ts-node` is not going to be a good fit. You might try tsx using @remagpie snippet above with this value: `nodeargs: --import tsx/esm`

Precisely, I had my eyes set on tsx also when suggesting a compositional approach.
Also thanks @remagpie

Pulumi will be brilliant to have broken up logically in an integrated nx monorepo, unfortunately its tsconfig has to extend the base one at the monorepo root. It has to be flexible enough to be able to use whatever is available in the codebase it slots in. The potential of type reuse and even business logic replicated in infrastructure is just too good.

**Comment by andyjy** [comment]
Incredible! Thanks @remagpie 

Tweaked to work with esm and ts-node using:

```yaml
runtime:
  name: nodejs
  options:
    typescript: false
    nodeargs: "--experimental-specifier-resolution=node --loader ts-node/esm"
```

**Comment by gunzy83** [comment]
Pulumi would likely have to support deno first https://github.com/pulumi/pulumi/issues/3838

**Comment by Stadly** [comment]
> Incredible! Thanks @remagpie
> 
> Tweaked to work with esm and ts-node using:
> 
> ```yaml
> runtime:
>   name: nodejs
>   options:
>     typescript: false
>     nodeargs: "--experimental-specifier-resolution=node --loader ts-node/esm"
> ```

There is a warning emitted when using `--loader ts-node/esm`:
```
(node:195555) ExperimentalWarning: `--experimental-loader` may be removed in the future; instead use `register()`:
--import 'data:text/javascript,import { register } from "node:module"; import { pathToFileURL } from "node:url"; register("ts-node/esm", pathToFileURL("./"));'
(Use `node --trace-warnings ...` to show where the warning was created)
```

To avoid it, you can use:

```yaml
runtime:
  name: nodejs
  options:
    typescript: false
    nodeargs: --experimental-specifier-resolution=node --import 'data:text/javascript,
      import { register } from "node:module";
      import { pathToFileURL } from "node:url";
      register("ts-node/esm", pathToFileURL(".").toString());
      '
```

**Comment by Shyam-Chen** [comment]
https://github.com/pulumi/pulumi/pull/15622

Pulumi v3.113.0

```sh
$ pnpm install ts-node typescript -D
```

**Comment by justinvp** [comment]
This has now been addressed with the release of `@pulumi/pulumi` v3.113.0. You can read more about it at https://www.pulumi.com/blog/typescript-versions/.

> With the release of the [Pulumi Node.js SDK 3.113.0](https://www.npmjs.com/package/@pulumi/pulumi) you can now use any version of TypeScript by adding it as a dependency to your project’s package.json file.
> 
> Pulumi supports all TypeScript versions from 3.8 and up, including [TypeScript 5.4](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-4.html).
> 
> If you have an existing project, you can upgrade your project to the latest versions of TypeScript and Pulumi by running:
> ```sh
> npm add typescript@^5.4.5 @pulumi/pulumi@^3.113.0
> ```
> 
> We have also updated our [Pulumi templates](https://github.com/pulumi/templates) to include the latest version of TypeScript. To create a new project, run:
> ```sh
> pulumi new
> ```
>
> Adding TypeScript as a dependency to your project is optional, as TypeScript and ts-node are optional peer dependencies of the Node.js SDK. This makes it easy to use a single version of TypeScript or ts-node within monorepos. To ensure backward compatibility, Pulumi ships with bundled versions of TypeScript 3.8 and ts-node 7, and falls back to these versions if they are not installed in your project’s node modules.

Note: TypeScript and ts-node are now optional peer dependencies of the Node.js SDK. This makes it possible to chose any version of these packages by adding them as dependency or devDependency to your package.json. If these optional peer dependencies are not installed, Pulumi will fallback to bundled versions included in the Node.js SDK to maintain backwards compatibility. As a side effect of this change, the `tsc` binary is no longer available in `node_modules/.bin/tsc`. If you rely on `tsc` being available in your project, add it as a dependency or devDependency to your package.json. Similarly, if you rely on `ts-node` being available in your project (e.g. to run tests like `npx jest`), add it as a dependency or devDependency to your package.json.
